### PR TITLE
support http/2 response trailers

### PR DESCRIPTION
**Pull request by sanlengjingvv**
#### Checklist

 - [ ] I have updated tests where applicable.
 - [ ] I have added an entry to the CHANGELOG.

### CODE REVIEW

**Review by mhils — CHANGES_REQUESTED** [review]
Hi @sanlengjingvv, thanks for the PR! I'm sorry that your previous HTTP/2 implementation got taken out when we switched to the new proxy core. I appreciate the effort here, but the implementation you are proposing here has the issue that trailers between concurrent flows may be mixed up. We need to define RequestTrailers/ResponseTrailers in [_events.py](https://github.com/mitmproxy/mitmproxy/blob/master/mitmproxy/proxy/layers/http/_events.py), emit them when receiving trailers, extend [HttpStream](https://github.com/mitmproxy/mitmproxy/blob/a00f247f0edfb303bd54cb1f0b7b4e97c4a56fd0/mitmproxy/proxy/layers/http/__init__.py#L93) to accomodate for that, and then add tests. I do have that on my radar, but it's admittedly not a high priority item at the moment (https://github.com/mitmproxy/mitmproxy/issues/4348). If you would like to give it a stab I'd be more than happy to help and explain things, feel free to ping me on Slack as well!

File: mitmproxy/proxy/layers/http/_http2.py [hunk 1]
```diff
@@ -448,6 +452,10 @@ def handle_h2_event(self, event: h2.events.Event) -> CommandGenerator[bool]:
             self.streams[event.stream_id] = StreamState.HEADERS_RECEIVED
             yield ReceiveHttp(ResponseHeaders(event.stream_id, response, bool(event.stream_ended)))
             return False
+        elif isinstance(event, h2.events.TrailersReceived):
+            pseudo_trailers, trailers = split_pseudo_headers(event.headers)
```
Comment (mhils) [inline mitmproxy/proxy/layers/http/_http2.py:456]: https://httpwg.org/specs/rfc7540.html#rfc.section.8.1.2.1:

> Pseudo-header fields MUST NOT appear in trailers. 

So we can skip that here :)

**Review by mhils — CHANGES_REQUESTED** [review]
This looks great, thanks for the PR! 🍰 Some comments below. 😃

**Review by mhils — COMMENTED** [review]
Thanks for the quick iteration! I have played around with this a bit further, added some streaming tests and then fixed a few streaming-related bugs. 

The most important change is that I have removed the requesttrailers/responsetrailers hooks. Instead, `request` and `response` now always fire after trailers have been received, but before they are forwarded. Can you test if things work for you as expected?

**Comment by sanlengjingvv** [comment]
Replaying a request with trailers will get an error
> Unexpected event during replay: <module 'mitmproxy.proxy.events' from '/Users/niconi/mitmproxy/mitmproxy/proxy/events.py'>

**Comment by mhils** [comment]
Fixed for me locally, does it work for you as well now? 😃

**Comment by sanlengjingvv** [comment]
> Fixed for me locally, does it work for you as well now? 😃

Yes, it works for me now.

**Review by mhils — APPROVED** [review]

**Comment by mhils** [comment]
Excellent. Thanks again for the initial PR and getting this going! 😃 🍰

**Pull request merged**